teps/0144-param-enum.md

```diff
@@ -0,0 +1,314 @@
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+  - [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Tasks](#tasks)
+    - [TaskRuns with TaskRef or Remote Resolution](#taskruns-with-taskref-or-remote-resolution)
+    - [TaskRuns with Embedded TaskSpec](#taskruns-with-embedded-taskspec)
+  - [Pipelines](#pipelines)
+    - [PipelineTasks with TaskRef or Remote Solution](#pipelinetasks-with-taskref-or-remote-solution)
+    - [Pipelines with Embedded PipelineTasks](#pipelines-with-embedded-pipelinetasks)
+    - [PipelineRuns](#pipelineruns)
+- [Design Evaluation](#design-evaluation)
+  - [Reusability](#reusability)
+  - [Simplicity](#simplicity)
+  - [Flexibility](#flexibility)
+  - [User Experience](#user-experience)
+  - [Performance](#performance)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Potential Future Work](#potential-future-work)
+<!-- /toc -->
+
+## Summary
+This TEP proposes adding the `enum` field for `string` param to the [`ParamSpec`](https://github.com/tektoncd/pipeline/blob/main/docs/pipeline-api.md#paramspec) in a Tekton [Task](https://github.com/tektoncd/pipeline/blob/main/docs/pipeline-api.md#taskspec) and [Pipeline](https://github.com/tektoncd/pipeline/blob/main/docs/pipeline-api.md#tekton.dev/v1.PipelineSpec). 
+
+This `enum` field will provide a way for Task and Pipeline authors to restrict a `string` parameter to a fixed set of values at authoring time. 
+
+## Motivation
+Parameters allow users to inject values to [certain fields](https://github.com/tektoncd/pipeline/blob/main/docs/variables.md#fields-that-accept-variable-substitutions) of their Pipeline/Task at execution time. 
+
+While it offers flexibility to the `Task`/`Pipeline` users, allowing any parameter injection to a field with no control has raised security concerns to `Task`/`Pipeline` authors since random parameter injection could unintentionally trigger a `Task` to do malicious things even though the `Task` content is signed and verified.
+
+Today, there is no standardized way to restrict parameter values in Tekton. `Task` authors may have to write an extra input validation `step`, or to embed the input validation logic in the scripts to achieve this validation against the user input. There are two problems with this:
+- **Usability issue**: Task authors are burdened with the task of implementing validation logic, which can lead to errors and inconsistencies.
+- **Error handling issue**: The errors caused by the invalid `param` values can only be caught at `pod` run time.
+
+Therefore, a built-in mechanism is desired to allow `Task`/`Pipeline` authors to declare upfront a list of allowed values for a parameter at authoring time. 
+
+In addition, this feature will enable Tekton to become a step closer to [SLSA hermeticity requirement](https://slsa.dev/spec/v0.1/requirements#hermetic), which requires that all inputs to a build must be fully declared up front. 
```

Comment:
I'm not sure I understand how this proposal helps in that sense, an ENUM is still a range of possible input values. The list of valid values is defined upfront but the actual value is not.

Reply:
Yeah, this is correct. If we have the mechanism to have the valid values declared upfront and fail (not running) all the invalid values, I think it meets the Hermetic requirement.

Reply:
I'm afraid I still do not understand your point but never mind, definitely not blocking